### ISSUE TITLE
Thread of applicaton insights agent runs in an endless loop, missing telemetry data

### ISSUE THREAD

**Issue by dvag-holger-krusemark** [microsoft/ApplicationInsights-Java#2218]
### Expected behavior
The application insights agent runs smoothly in the background without significant impact on the server or the application.

### Actual behavior
In rare cases, one of the "BatchSpanProcessor_WorkerThread-1" thread runs in an endless loop and occupies an entire cpu core.

Stacktrace of this thread:

java.lang.Thread.yield(Native Method)
java.util.concurrent.locks.AbstractQueuedSynchronizer.transferAfterCancelledWait(AbstractQueuedSynchronizer.java:1752)
java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2119)
java.util.concurrent.ArrayBlockingQueue.poll(ArrayBlockingQueue.java:432)
io.opentelemetry.sdk.trace.export.BatchSpanProcessor$Worker.run(BatchSpanProcessor.java:230)
java.lang.Thread.run(Thread.java:829)

When the problem started, some telemetry data was missing in the Application Insight portal.

### To Reproduce
The problem cannot be reproduced specifically. It occurred twice in a load and performance test at high load and once in a productive system under normal load.

#### Sample Application
-

### System information
Please provide the following information:
 - SDK Version: 3.2.4
 - OS type and version: CentOS Linux 7 (runs in a VM from VMWare), Ubuntu 20.04.4 LTS (runs in a pod on a VM in azure)
 - Application Server type and version (if applicable): Tomcat, spring-boot
 - Using spring-boot? yes
 - Additional relevant libraries (with version, if applicable): -
 
### Logs
No log files, but a heap dump of the Java VM (including the state of all Java objects) is available. For data protection reasons, however, I cannot publish this dump. However, I will be happy to answer any questions about it.

### Screenshots
-

**Comment by trask** [comment]
hi @dvag-holger-krusemark! this sounds like #2112, can you try with env var `APPLICATIONINSIGHTS_PREVIEW_BSP_SCHEDULE_DELAY=100` and see if that resolves the issue? if it resolves the issue we will probably make that the default, just not sure how to reproduce this, so need some help, thx!

**Comment by trask** [comment]
"BatchSpanProcessor_WorkerThread-1" thread is expected to run in an "endless" loop, but it should be spending most of its time sleeping, waiting on a signal that there is new telemetry to process. I'm suspecting that by not having any "delay" (the current default), in some environments this is causing it to wake up too often and, I'm really unsure though since we have not been able to reproduce this issue, so it's a guess until we are able to hear back from you on the test results

**Comment by dvag-holger-krusemark** [comment]
The delay inside the infinite loop of the consumer thread is 5ms (in the heapdump I see that the attribute "scheduleDelayNanos" contained the value 5000000000).
The real problem is that the thread from the call "java.util.concurrent.ArrayBlockingQueue.poll()" does not return. The thread is trapped in the infinite loop in java.util.concurrent.locks.AbstractQueuedSynchronizer.transferAfterCancelledWait() and keeps calling Thread.yield().

Therefore, I don't think APPLICATIONINSIGHTS_PREVIEW_BSP_SCHEDULE_DELAY=100 fundamentally solves the problem.

Using the queue as a notification mechanism can lead to another severe problem (see https://github.com/microsoft/ApplicationInsights-Java/issues/2056). After testing with the latest Java 11 version (11.0.14.1): an instance of java.util.concurrent.ArrayBlockingQueue can still get into an inconsistent state when a StackOverflowError occurs -> The lock is not released.

-> Using java.util.concurrent.ArrayBlockingQueue in "BatchSpanProcessor_Worker" can lead to two severe problems affecting the application.

We wanted to update to 3.2.10. However, I'm afraid that we will get a new problem with https://github.com/microsoft/ApplicationInsights-Java/issues/2219.
Should we wait for a fix for this bug?

**Comment by trask** [comment]
> The real problem is that the thread from the call "java.util.concurrent.ArrayBlockingQueue.poll()" does not return. The thread is trapped in the infinite loop in java.util.concurrent.locks.AbstractQueuedSynchronizer.transferAfterCancelledWait() and keeps calling Thread.yield().

how are you detecting that this is causing high CPU? (as opposed to a harmless wait)

I wonder if the comment at https://github.com/microsoft/ApplicationInsights-Java/issues/2112#issuecomment-1095583259 applies here as well?

For the missing telemetry data issue, 3.2.10 has significant improvement:

> In a simple local logging test, I can get ~30x(!) more throughput without dropping any telemetry (still using the single thread, just not blocking on responses from the ingestion service).

(https://github.com/microsoft/ApplicationInsights-Java/issues/2129#issuecomment-1066198026)

> We wanted to update to 3.2.10. However, I'm afraid that we will get a new problem with https://github.com/microsoft/ApplicationInsights-Java/issues/2219.
Should we wait for a fix for this bug?

Have you configured HTTP Headers to be captured in your json configuration? If not then that issue shouldn't affect you. If you have, and you are able to reproduce that issue, any details to help us reproduce it would help us to fix it.

**Comment by dvag-holger-krusemark** [comment]
To make sure that the mentioned infinite loop is responsible for the high CPU load I have
1. created many thread dumps to see where the thread hangs
2. performed a code review, determined the conditions for the infinite loop via analysis of the heapdump and reimplemented the loop.

In https://github.com/microsoft/ApplicationInsights-Java/issues/2219 you can also see the java.util.concurrent.ArrayBlockingQueue.poll() method in the screenshot.

I suspect that by blocking the thread, the transmission of telemetry data is disrupted.

We have not configured recording of individual http headers.
However, there was a serious issue recently with version 3.2.10 in io.opentelemetry.sdk.metrics.internal.aggregator.DoubleHistogramAggregator. Hundreds of threads were blocked by an unresolvable lock (java.util.concurrent.locks.ReentrantLock). Analyzing the heap dump, I noticed that stacks of the threads was very deep (up to a call depth of 11,500, code from package "io.opentelemetry.javaagent" and  "sun.net.www.protocol.http.HttpURLConnection").
This is a  dangerous because: ReentrantLock + StackoverflowError + Java 11 -> unresolvable lock (see https://github.com/microsoft/ApplicationInsights-Java/issues/2056, but this time in a different place).
-> New Issue in 3.2.10?

**Comment by trask** [comment]
> there was a serious issue recently with version 3.2.10 in io.opentelemetry.sdk.metrics.internal.aggregator.DoubleHistogramAggregator. Hundreds of threads were blocked by an unresolvable lock (java.util.concurrent.locks.ReentrantLock)

Can you attach thread dumps for these? Thx!

**Comment by dvag-holger-krusemark** [comment]
The file "deep call stack.zip" contains the stack trace during instrumentation.

Because of the risks we stay on version 3.2.4.
I suspect that "APPLICATIONINSIGHTS_PREVIEW_BSP_SCHEDULE_DELAY=100" will not change the problem "endless loop consumes entire cpu core".

**Comment by dvag-holger-krusemark** [comment]
[deep call stack.zip](https://github.com/microsoft/ApplicationInsights-Java/files/8481187/deep.call.stack.zip)

**Comment by trask** [comment]
thanks @dvag-holger-krusemark!

this looks like the same issue as #2219

are you able to reproduce it? if so, can you try with the latest SNAPSHOT build and see if that resolves the issue or not?

https://github.com/microsoft/ApplicationInsights-Java/suites/5977750973/artifacts/205795109

or, can you share how you are using HttpURLConnection, so that I can try to construct a repro of the issue?

**Comment by dvag-holger-krusemark** [comment]
The problem has so far only occurred in a productive environment. However, I cannot import a snapshot version there.

Our code only called "sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream()". 

What is the current status of the other issues?

1. Thread "BatchSpanProcessor_WorkerThread-1" is stuck in an endless loop in "java.util.concurrent.locks.AbstractQueuedSynchronizer.transferAfterCancelledWait".

2. Stackoverflow in ReentrantLock leads to unreleasable lock (-> https://github.com/microsoft/ApplicationInsights-Java/issues/2056). I have seen that the ticket has been closed. However, from my point of view, the problem has not been solved. This scenario can be reproduced with a test programm.
Java should, of course, ensure that ReentrantLock does not break during a stack overflow. But there should also be no stack overflow in the first place.

**Comment by trask** [comment]
> This scenario can be reproduced with a test programm.

can you share your test program?

also, it looks like the upstream changes (most likely https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/5673) may have solved #2219

**Comment by dvag-holger-krusemark** [comment]
The test program "QueueTest" exchanges data between threads (five producer threads and one consumer thread) via a queue like "BatchSpanProcessor".
To create a stack overflow in the producer threads, the call depth is increased step by step until the queue is accessed. When a stack overflow occurs, the producer thread terminates.
The main thread monitors the throughput of the data exchange and identifies the cause based on known patterns when it freezes.

Meanwhile I can reproduce both problems with the test program under Windows 10/Java-11 (Core i7, 4 cores/8 treads).

1. "Thread is stuck in an endless loop in "java.util.concurrent.locks.AbstractQueuedSynchronizer.transferAfterCancelledWait calling Thread.yield()."
For this scenario to occur reliably, the Java VM's JIT compiler must be disabled ("-Djava.compiler=NONE"). The Java VM will then also detect the problem
and displays the warning message
"OpenJDK 64-bit Server VM warning: Potentially dangerous stack overflow in ReservedStackAccess annotated method java.util.concurrent.locks.ReentrantLock$Sync.nonfairTryAcquire(I)Z"
-> The main thread then reports "Endless loop in yield()".

2. "Stack overflow in ReentrantLock leads to unreleasable lock."
This problem only occurs reliably when the Java VM's JIT compiler is enabled.
-> The main thread then reports "Lock owning producer thread is dead."

-> Both problems are the result of a stack overflow when accessing ArrayBlockingQueue/ReentrantLock.

**Comment by dvag-holger-krusemark** [comment]
[QueueTest.zip](https://github.com/microsoft/ApplicationInsights-Java/files/8564396/QueueTest.zip)

**Comment by dvag-holger-krusemark** [comment]
So far I know only one cause for stack overflow: https://github.com/microsoft/ApplicationInsights-Java/issues/2219
According to my observations, this problem can apparently also occur in version 3.2.4.
When and in which release will this bug be fixed?
Are there any other scenarios where stack overflow can occur?

**Comment by floriankoch** [comment]
@trask   any progress here?

**Comment by trask** [comment]
hi @floriankoch, can you set `APPLICATIONINSIGHTS_PREVIEW_BSP_SCHEDULE_DELAY` environment variable to `5000`?

there is one downside to that, which is live metrics will be delayed by up to 5 seconds, but if it resolves the production cpu utilization issue, then we will know what the underlying issue is, and we will work on fixing it in a way that does not impact live metrics

btw, please follow also #2112, I believe it is the same issue

**Comment by dvag-holger-krusemark** [comment]
@trask: The cause of the listed infinite loop is the result of an inconsistent state caused by a stack overflow error.
Therefore, I believe that changing APPLICATIONINSIGHTS_PREVIEW_BSP_SCHEDULE_DELAY to 5000  is not a clean solution.
Were you able to recreate the problem with the test program?

**Comment by trask** [comment]
hi @dvag-holger-krusemark, are you still seeing this issue, or ok to close?

**Comment by floriankoch** [comment]
@trask  we wait for the 3.3.0 Relase and test again

**Comment by trask** [comment]
@dvag-holger-krusemark 3.3.0 is released, looking forward to your test results, thx!
